### PR TITLE
Feature/vvolkov/dev conditional compilation

### PR DESCRIPTION
**Pull request by vladislav-volkov**

### CODE REVIEW

File: src/cpu/cpu_concat.cpp [hunk 1]
```diff
@@ -18,25 +18,48 @@
...
+            dnnl::impl::engine_t *,
+            const dnnl::impl::primitive_attr_t *,
+            const dnnl::impl::memory_desc_t *,
+            int,
+            int,
+            const dnnl::impl::memory_desc_t *);
+
+# define CPD_INSTANCE(...) REG_MKLDNN_FN(cpd_builder, __VA_ARGS__)
+
+#else   // SELECTIVE_BUILD == ON || SELECTIVE_BUILD == OFF
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/cpu/cpu_concat.cpp:46]: Here it means that this block is enabled if SELECTIVE_BUILD_ANALYZER is not defined. I'll change this strange comment. (:

File: src/common/memory_zero_pad.cpp [hunk 2]
```diff
@@ -21,9 +21,11 @@
 #include "stream.hpp"
 #include "type_helpers.hpp"
 #include "utils.hpp"
+#include "mkldnn_sel_build.hpp"
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/common/memory_zero_pad.cpp]: fixed

File: src/common/mkldnn_macros.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*******************************************************************************
+* Copyright 2016-2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/common/mkldnn_macros.hpp]: fixed

File: src/common/mkldnn_sel_build.hpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/common/mkldnn_sel_build.hpp]: merged into one dnnl_sel_build.hpp

File: src/cpu/cpu_concat.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/cpu/cpu_concat.cpp]: fixed

File: src/cpu/cpu_convolution_list.cpp [hunk 5]
```diff
@@ -141,33 +141,33 @@ const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
...
+        CPU_INSTANCE(ref_convolution_bwd_data_t, f32, f32, f32, f32)
         nullptr,
     }},
     {{backward_data, f32, bf16, bf16}, {
-        CPU_INSTANCE_X64(jit_uni_dw_convolution_bwd_data_t<avx512_core, bf16, f32>)
-        CPU_INSTANCE_X64(jit_uni_fork_dw_convolution_bwd_data_t<avx512_core, bf16, f32>)
-        CPU_INSTANCE_X64(jit_avx512_core_bf16_1x1_convolution_bwd_data_t<f32>)
+        CPU_INSTANCE_X64(jit_uni_dw_convolution_bwd_data_t, avx512_core, bf16, f32)
+        CPU_INSTANCE_X64(jit_uni_fork_dw_convolution_bwd_data_t, avx512_core, bf16, f32)
+	CPU_INSTANCE_X64(jit_avx512_core_bf16_1x1_convolution_bwd_data_t, f32)
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/cpu/cpu_convolution_list.cpp]: fixed

File: src/cpu/cpu_softmax_list.cpp [hunk 6]
```diff
@@ -35,17 +35,17 @@ using namespace dnnl::impl::data_type;
...
-        CPU_INSTANCE(ref_softmax_fwd_t<bf16>)
-        CPU_INSTANCE(ref_softmax_bwd_t<bf16>)
+        CPU_INSTANCE_X64(jit_uni_softmax_fwd_t, avx512_common)
+        CPU_INSTANCE_X64(jit_uni_softmax_bwd_t, avx512_common)
+        CPU_INSTANCE_X64(jit_uni_softmax_fwd_t, avx2)
+        CPU_INSTANCE_X64(jit_uni_softmax_fwd_t, sse41)
+        CPU_INSTANCE_X64(jit_uni_fork_softmax_fwd_t, avx512_common)
+        CPU_INSTANCE_X64(jit_uni_fork_softmax_fwd_t, avx2)
+        CPU_INSTANCE_X64(jit_uni_fork_softmax_fwd_t, sse41)
+	CPU_INSTANCE(ref_softmax_fwd_t, f32)
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/cpu/cpu_softmax_list.cpp]: fixed

File: src/common/memory_zero_pad.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmitry-gorokhov) [reply src/common/memory_zero_pad.cpp]: I meant not only file names but all other stuff as well (like macros name: MKLDNN_MACRO_* -> DNNL_MACRO_*)

[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply src/common/memory_zero_pad.cpp]: ok, fixed

**Pull request merged**